pumpkin-py/src/model.rs

```diff
@@ -0,0 +1,185 @@
+use std::num::NonZero;
+use std::path::PathBuf;
+
+use pumpkin_lib::containers::KeyedVec;
+use pumpkin_lib::options::LearningOptions;
+use pumpkin_lib::options::SolverOptions;
+use pumpkin_lib::proof::Format;
+use pumpkin_lib::proof::ProofLog;
+use pumpkin_lib::termination::Indefinite;
+use pumpkin_lib::ConstraintOperationError;
+use pumpkin_lib::Solver;
+use pyo3::prelude::*;
+
+use crate::constraints::Constraint;
+use crate::result::SatisfactionResult;
+use crate::result::Solution;
+use crate::variables::BoolVariable;
+use crate::variables::IntExpression;
+use crate::variables::IntVariable;
+use crate::variables::VariableMap;
+
+#[pyclass]
+#[derive(Default)]
+pub struct Model {
+    integer_variables: KeyedVec<IntVariable, ModelIntVar>,
+    boolean_variables: KeyedVec<BoolVariable, Option<String>>,
+    constraints: Vec<ModelConstraint>,
+}
+
+#[pymethods]
+impl Model {
+    #[new]
+    fn new() -> Model {
+        Model::default()
+    }
+
+    /// Create a new integer variable.
+    #[pyo3(signature = (lower_bound, upper_bound, name=None))]
+    fn new_integer_variable(
+        &mut self,
+        lower_bound: i32,
+        upper_bound: i32,
+        name: Option<&str>,
+    ) -> IntExpression {
+        let variable = ModelIntVar {
+            lower_bound,
+            upper_bound,
+            name: name.map(|n| n.to_owned()),
+        };
+
+        self.integer_variables.push(variable).into()
+    }
+
+    /// Create a new boolean variable.
+    #[pyo3(signature = (name=None))]
+    fn new_boolean(&mut self, name: Option<&str>) -> BoolVariable {
+        self.boolean_variables.push(name.map(|n| n.to_owned()))
+    }
+
+    /// Add the given constraint to the model.
+    #[pyo3(signature = (constraint, tag=None))]
+    fn add_constraint(&mut self, constraint: Constraint, tag: Option<NonZero<u32>>) {
+        self.constraints.push(ModelConstraint {
+            constraint,
+            premise: None,
+            tag,
+        });
+    }
+
+    /// Add `implied_by -> constraint` to the model.
+    #[pyo3(signature = (constraint, premise, tag=None))]
+    fn add_reified(
```

Comment:
I dislike that we have the method `implied_by` in the constraints interface of pumpkin-lib which means half-reification while here `add_reified` means half-reification; it would be good to be consistent with the naming